### PR TITLE
feat: Add (optional) Jacobian reversal to Track

### PR DESCRIPTION
**Pull request by paulgessinger**
`TrackProxy::reverseTrackStates` can now optionally invert its jacobians, and move them to the correct corresponding track state.

### CODE REVIEW

**Review by andiwand — APPROVED** [review]

**Comment by paulgessinger** [comment]
Compiler segfaults again, maybe due to zip and track state ranges?

File: Core/include/Acts/EventData/TrackProxy.hpp [hunk 1]
```diff
@@ -673,19 +673,38 @@ class TrackProxy {
...
     IndexType current = tipIndex();
     IndexType next = kInvalid;
     IndexType prev = kInvalid;
 
     stemIndex() = tipIndex();
 
+    auto getTrackState = [this](IndexType idx) {
+      return m_container->trackStateContainer().getTrackState(idx);
+    };
+
```
Comment (paulgessinger) [inline Core/include/Acts/EventData/TrackProxy.hpp:688]:
```suggestion
```

File: Core/include/Acts/EventData/TrackProxy.hpp [hunk 2]
```diff
@@ -673,19 +673,38 @@ class TrackProxy {
...
+    // @TODO: Maybe refactor to not need this variable if invertJacobians == false
+    BoundMatrix nextJ;
+
     while (current != kInvalid) {
-      auto ts = m_container->trackStateContainer().getTrackState(current);
+      auto ts = getTrackState(current);
```
Comment (paulgessinger) [inline Core/include/Acts/EventData/TrackProxy.hpp:693]:
```suggestion
      auto ts = m_container->trackStateContainer().getTrackState(current);
```

File: Core/include/Acts/EventData/TrackProxy.hpp [hunk 3]
```diff
@@ -673,19 +673,33 @@ class TrackProxy {
...
       ts.template component<IndexType>(hashString("next")) = prev;
       ts.previous() = next;
+      if (invertJacobians) {
+        if (next != kInvalid) {
+          BoundMatrix curJ = ts.jacobian();
+          ts.jacobian() = nextJ.inverse();
+          nextJ = curJ;
+        } else {
+          nextJ = ts.jacobian();
+          ts.jacobian().setZero();
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/EventData/TrackProxy.hpp:700]: Not that I know of. I can probably manually assign `nan` but not sure that makes too much sense either.

[the remark this answers is not in the document]
Comment (andiwand) [reply Core/include/Acts/EventData/TrackProxy.hpp:700]: I would also prefer 0 over nan. The alternative would be to make the jacobian optional with `std::optional` or in the container but I think the overhead is not too bad and not having one on the first state is a question of definition / documentation

File: Tests/UnitTests/Core/EventData/TrackTests.cpp [hunk 4]
```diff
@@ -28,6 +28,7 @@
 #include "Acts/Utilities/Holders.hpp"
 #include "Acts/Utilities/Zip.hpp"
 
+#include <algorithm>
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Tests/UnitTests/Core/EventData/TrackTests.cpp:31]: Yeah no 😅

**Comment by paulgessinger** [comment]
All changed @felix-russo.

**Review by felix-russo — APPROVED** [review]
Good to go from my side (except if you want to add some documentation about the last/first Jacobian that you set to 0)

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [6df0704dd072984d1635c180954a0fb905fc3a19]
### :white_check_mark: All tests successful

|status|job|
|------|---|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33429912) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsWorkflow.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33429911) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33429910) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsCoreSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33429909) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33429908) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33429907) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateOrthogonalSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33429906) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateClusters.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33429905) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsPersistifyEDM.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33429904) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsKfRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33429903) |
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsExtrapolationAlgTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33429902) |
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsITkTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33429901) |
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33429900) |
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33429899) |
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33429898) |
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33429897) |

**Pull request merged**